### PR TITLE
Group by a categorical Series of unequal length

### LINKED ISSUE #44179: BUG: Should be able to group by a categorical Series of unequal length
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd
series = pd.Series(range(7))
grouper = pd.Series(list("ABBA"), dtype="category")
series.groupby(grouper)

# Raises ValueError: Length of grouper (4) and axis (7) must be same length
```

### Issue Description

When grouping by a `Series`, the values of the series after alignment determine the groups. In the example, however, pandas does not attempt to align the series and instead unnecessarily wants it to have the same length as the axis of grouping.

### Expected Behavior

I expect `obj.groupby(grouper)`, where `grouper` is a `Series` with a categorical data type, to allow the grouper to have unequal length and to align it, as in the following where the grouper has a non-categorical data type:

```python
series.groupby(grouper.astype(object)).groups
# Returns {'A': [0, 3], 'B': [1, 2]}
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 9b2bb732f09d8e870d806b337b97d3d99fd5b118
python           : 3.8.12.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.17134
machine          : AMD64
processor        : Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : English_United States.1252

pandas           : 1.4.0.dev0+952.g9b2bb732f0
numpy            : 1.21.2
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 58.0.4
Cython           : 0.29.24
pytest           : 6.2.5
hypothesis       : 6.23.1
sphinx           : 4.2.0
blosc            : None
feather          : None
xlsxwriter       : 3.0.1
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.28.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : 1.3.2
fsspec           : 2021.10.0
fastparquet      : 0.7.1
gcsfs            : 2021.10.0
matplotlib       : 3.4.3
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 5.0.0
pyxlsb           : None
s3fs             : 2021.10.0
scipy            : 1.7.1
sqlalchemy       : 1.4.25
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by stanwest**
- [X] closes #44179
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [X] whatsnew entry

While this fix enables grouping by a `Series` that has a categorical data type and length unequal to the axis of grouping, it maintains the requirement (developed in #3017, #9741, and #18525) that the length of a `Categorical` object used for grouping must match that of the axis of grouping. Additionally, this fix checks that requirement consistently with other groupers whose lengths must match—namely `list`, `tuple`, `Index`, and `numpy.ndarray`—and produces the same exception message when the lengths differ.

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
cc @rhshadrach if any comments

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -593,7 +593,7 @@ Groupby/resample/rolling
 - Fixed bug in :meth:`Series.rolling` and :meth:`DataFrame.rolling` for centered datetimelike windows with uneven nanosecond (:issue:`43997`)
 - Bug in :meth:`GroupBy.nth` failing on ``axis=1`` (:issue:`43926`)
 - Fixed bug in :meth:`Series.rolling` and :meth:`DataFrame.rolling` not respecting right bound on centered datetime-like windows, if the index contain duplicates (:issue:`#3944`)
-
+- Enabled grouping by a :class:`Series` that has a categorical data type and length unequal to the axis of grouping (:issue:`44179`); also, with this fix, the ``ValueError`` raised when grouping by a :class:`Categorical` with unequal length now has the same message as when grouping by other sequences
```
Comment (rhshadrach) [inline doc/source/whatsnew/v1.4.0.rst:596]: We typically don't include changes to the error messages themselves; only when the type of the error (e.g. ValueError vs TypeError) changes. Can you remove the 2nd clause.

Comment (rhshadrach) [reply doc/source/whatsnew/v1.4.0.rst:596]: Also, this is worded as an enhancement - can you reword as a bugfix "Fixed bug when grouping by..."

File: pandas/tests/groupby/test_categorical.py [hunk 2]
```diff
@@ -668,11 +668,32 @@ def test_bins_unequal_len():
     bins = pd.cut(series.dropna().values, 4)
 
     # len(bins) != len(series) here
-    msg = r"Length of grouper \(8\) and axis \(10\) must be same length"
-    with pytest.raises(ValueError, match=msg):
+    with pytest.raises(ValueError, match=r"Grouper and axis must be same length"):
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_categorical.py:671]: No need for `r` since the parentheses were removed.

File: pandas/tests/groupby/test_categorical.py [hunk 3]
```diff
@@ -668,11 +668,32 @@ def test_bins_unequal_len():
...
+        (Series(range(7)), {"A": [0, 3], "B": [1, 2]}),
+    ],
+)
+def test_categorical_series(series, data):
+    # Group the given series by a series with categorical data type such that group A
+    # takes indices 0 and 3 and group B indices 1 and 2, obtaining the values mapped in
+    # the given data.
+    groupby = series.groupby(Series(list("ABBA"), dtype="category"))
+    tm.assert_series_equal(
+        groupby.aggregate(list), Series(data, index=CategoricalIndex(data.keys()))
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_categorical.py:693]: Can you separate out result and expected to their own lines:

```
result = groupby.aggregate(list)
expected = Series(data, index=CategoricalIndex(data.keys())
tm.assert_series_equal(result, expected)
```

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR! Generally looks good - some minor requests below. Also, I think we should add a test where `.groupby(..., dropna=False)`.

**Comment by stanwest** [comment]
> Thanks for the PR! Generally looks good - some minor requests below.

Thanks for the appreciation! I believe that I've addressed the minor requests.

> Also, I think we should add a test where `.groupby(..., dropna=False)`.

Although I'm open to hearing more about what you have in mind, to me the ``dropna`` functionality seems beyond the scope of the bug and fix. That is, the bug and fix concern whether the ``groupby`` method either rejects a ``Series`` grouper solely because it has a categorical data type or accepts and reindexes it, and that precedes forming groups and deciding whether to drop any.

**Comment by rhshadrach** [comment]
@stanwest - I believe the bugfix you've done here results in new cases (since they would raise before) where the grouper has null values. It seems prudent to me to test these cases.

**Comment by jreback** [comment]
> @stanwest - I believe the bugfix you've done here results in new cases (since they would raise before) where the grouper has null values (e.g. [#44179 (comment)](https://github.com/pandas-dev/pandas/issues/44179#issuecomment-950980759)). It seems prudent to me to test these cases.

agree

**Comment by jreback** [comment]
can you merge master and address the open comment

**Comment by stanwest** [comment]
> @stanwest - I believe the bugfix you've done here results in new cases (since they would raise before) where the grouper has null values (e.g. [#44179 (comment)](https://github.com/pandas-dev/pandas/issues/44179#issuecomment-950980759)). It seems prudent to me to test these cases.

I might otherwise agree, but I've found the behavior of *pandas* in related cases sometimes to be inconsistent with the documentation or otherwise surprising.  Consider the following demonstration:

```python
import pandas as pd

pd.options.display.max_colwidth = None

def outcome(dtype, uneq_len, uneq_index, dropna, operation):
    """Return the result or exception from a groupby operation."""
    ser = pd.Series(range(6))
    grouper = pd.Series([*"ABBA"] + (3 if uneq_len else 2) * [pd.NA], dtype=dtype)
    if uneq_index:
        grouper = grouper.reindex(grouper.index - 1)
    try:
        val = operation(ser.groupby(grouper, dropna=dropna))
    except ValueError as exc:
        val = exc
    return val

def summary(obj):
    """Return a summary representation of the given object."""
    if isinstance(obj, pd.Series):
        return (
            f"{type(obj).__name__}"
            f"({obj.to_dict()}, index={type(obj.index).__name__}(...))"
        )
    else:
        return repr(obj)

pd.Series(
    {
        (oper_name, dtype, dropna, uneq_len, uneq_index): summary(
            outcome(dtype, uneq_len, uneq_index, dropna, operation)
        )
        for oper_name, operation in {
            "groups": lambda gby: gby.groups,
            "agg(list)": lambda gby: gby.aggregate(list),
        }.items()
        for dtype in ("object", "category")
        for dropna in (False, True)
        for uneq_len in (False, True)
        for uneq_index in (False, True)
    },
).rename_axis(index=["operation", "dtype", "dropna", "uneq_len", "uneq_index"])
```

It groups the same series by another series of `object` or categorical data type, equal or unequal length, and equal or unequal index, either keeping or dropping missing keys, and then either produces the `groups` attribute or aggregates each group with `list` and represents the result or exception.  It constructs the grouper such that group "A" should obtain 0 and 3; group "B", 1 and 2; and any group for missing values in the grouper, 4 and 5.  (Although `groups` obtains the indices and `aggregate` obtains the values, the two are equal for the series being grouped to aid inspection.)  With the present PR applied, it produces the following output, the rows of which I've manually numbered:

```
    operation  dtype     dropna  uneq_len  uneq_index
 1  groups     object    False   False     False                       ValueError('Categorical categories cannot be null')
 2                                         True                        ValueError('Categorical categories cannot be null')
 3                               True      False                       ValueError('Categorical categories cannot be null')
 4                                         True                        ValueError('Categorical categories cannot be null')
 5                       True    False     False                                                {'A': [0, 3], 'B': [1, 2]}
 6                                         True                                                 {'A': [0, 3], 'B': [1, 2]}
 7                               True      False                                                {'A': [0, 3], 'B': [1, 2]}
 8                                         True                                                 {'A': [0, 3], 'B': [1, 2]}
 9             category  False   False     False                                                {'A': [0, 3], 'B': [1, 2]}
10                                         True                                                 {'A': [0, 3], 'B': [1, 2]}
11                               True      False                                                {'A': [0, 3], 'B': [1, 2]}
12                                         True                                                 {'A': [0, 3], 'B': [1, 2]}
13                       True    False     False                                                {'A': [0, 3], 'B': [1, 2]}
14                                         True                                                 {'A': [0, 3], 'B': [1, 2]}
15                               True      False                                                {'A': [0, 3], 'B': [1, 2]}
16                                         True                                                 {'A': [0, 3], 'B': [1, 2]}
17  agg(list)  object    False   False     False         Series({'A': [0, 3], 'B': [1, 2], nan: [4, 5]}, index=Index(...))
18                                         True          Series({'A': [0, 3], 'B': [1, 2], nan: [4, 5]}, index=Index(...))
19                               True      False         Series({'A': [0, 3], 'B': [1, 2], nan: [4, 5]}, index=Index(...))
20                                         True          Series({'A': [0, 3], 'B': [1, 2], nan: [4, 5]}, index=Index(...))
21                       True    False     False                      Series({'A': [0, 3], 'B': [1, 2]}, index=Index(...))
22                                         True                       Series({'A': [0, 3], 'B': [1, 2]}, index=Index(...))
23                               True      False                      Series({'A': [0, 3], 'B': [1, 2]}, index=Index(...))
24                                         True                       Series({'A': [0, 3], 'B': [1, 2]}, index=Index(...))
25             category  False   False     False           Series({'A': [0, 3], 'B': [1, 2]}, index=CategoricalIndex(...))
26                                         True            Series({'A': [0, 3], 'B': [1, 2]}, index=CategoricalIndex(...))
27                               True      False           Series({'A': [0, 3], 'B': [1, 2]}, index=CategoricalIndex(...))
28                                         True            Series({'A': [0, 3], 'B': [1, 2]}, index=CategoricalIndex(...))
29                       True    False     False           Series({'A': [0, 3], 'B': [1, 2]}, index=CategoricalIndex(...))
30                                         True            Series({'A': [0, 3], 'B': [1, 2]}, index=CategoricalIndex(...))
31                               True      False           Series({'A': [0, 3], 'B': [1, 2]}, index=CategoricalIndex(...))
32                                         True            Series({'A': [0, 3], 'B': [1, 2]}, index=CategoricalIndex(...))
```

The present PR affects the cases with categorical data type and unequal length (namely rows 11, 12, 15, 16, 27, 28, 31, and 32): Instead of raising `ValueError('Length of grouper (7) and axis (6) must be same length')`, those cases return the same output as when the lengths are equal.

The cases with `dropna=False` seem not entirely correct.  For the `groups` operation, the cases with `object` data type (rows 1–4) suffer from issue #35202, and those with "category" data type (rows 9–12) lack an item for missing values.  I expected a result of `{'A': [0, 3], 'B': [1, 2], nan: [4, 5]}`.  A comparable case using two identical groupings includes an item for missing values:

```python
In [2]: pd.Series(range(6)).groupby(2 * [pd.Series([*"ABBA"] + 2 * [pd.NA], dtype=None)], dropna=False).groups
Out[2]: {('A', 'A'): [0, 3], ('B', 'B'): [1, 2], (nan, nan): [4, 5]}
```

Interestingly, the result is the same with `dropna=True`, as though the `groups` attribute is to express the groups prior to any filtering due to the `dropna` argument.  If that is the case, then rows 1–16 should produce `{'A': [0, 3], 'B': [1, 2], nan: [4, 5]}`.

For the `agg(list)` operation, the cases with `object` data type (rows 17–20) produce what I expect, but I expected those with "category" data type (rows 25–28) to include an item for missing values (`Series({'A': [0, 3], 'B': [1, 2], nan: [4, 5]}, index=CategoricalIndex(...))`).

Additionally, and perhaps related to the above mix of behaviors, I've found no `groupby` tests that cover both grouping by categorical data types and `dropna=False`.

To summarize my view, while some behaviors were questionable before this PR and remain so, this PR removes an unnecessary `ValueError` and makes the behaviors more consistent.

File: pandas/core/groupby/grouper.py [hunk 4]
```diff
@@ -938,7 +932,7 @@ def _convert_grouper(axis: Index, grouper):
             return grouper.reindex(axis)._values
     elif isinstance(grouper, MultiIndex):
         return grouper._values
-    elif isinstance(grouper, (list, tuple, Series, Index, np.ndarray)):
+    elif isinstance(grouper, (list, tuple, Index, Categorical, np.ndarray)):
```
Comment (jreback) [inline pandas/core/groupby/grouper.py:935]: why is this changed?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/grouper.py (hunk 4 above)
Comment (stanwest) [reply pandas/core/groupby/grouper.py:935]: First, `Series` has no effect, because a prior `elif` [clause](https://github.com/pandas-dev/pandas/blob/8d10ab4facf7225862150ca7695f60f8f5402bfe/pandas/core/groupby/grouper.py#L934) catches instances of `Series`.  The [first commit](https://github.com/pandas-dev/pandas/pull/44180/commits/8d10ab4facf7225862150ca7695f60f8f5402bfe) of this PR removes `Series` to clean up the line before other changes.

Second, inserting `Categorical` replaces and tightens the check formerly performed by the following [line](https://github.com/pandas-dev/pandas/blob/8d10ab4facf7225862150ca7695f60f8f5402bfe/pandas/core/groupby/grouper.py#L890) in `get_grouper()`:

```python
if is_categorical_dtype(gpr) and len(gpr) != obj.shape[axis]:
```

Removing that `if` block and inserting `Categorical` into the `isinstance` check is the crux of the bug fix.  By removing the `if` block, we in particular allow `Series` groupers with categorical data type and length unequal to that of the grouping axis to proceed and eventually be reindexed.  By inserting `Categorical`, we continue to require that groupers of that type (which are array-like and lack an index) have the same length as the axis of grouping.

**Comment by rhshadrach** [comment]
Thanks for the detailed response. I agree this PR moves in the right direction, and from your post the following two examples

```
gb = pd.Series(range(2)).groupby([pd.Series(["A", pd.NA], dtype="category")], dropna=False)
print(gb.size())

gb = pd.Series(range(2)).groupby([pd.Series(["A", np.nan], dtype="category")], dropna=False)
print(gb.size())
```

already produce the incorrect result (drop the null group). Is there an issue for this? I'm +1 on the behavior here.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
ok i think this is fine, @rhshadrach your last comment i think we should open a new issue but not block here right?

**Comment by stanwest** [comment]
> Is there an issue for this?

Yes: In issue #36327, grouping with `dropna=False` by a `Categorical` (as in the original comment of that issue) or by a column with categorical data type (as in [a follow-up comment](https://github.com/pandas-dev/pandas/issues/36327#issuecomment-728979374)) incorrectly drops the null groups. I believe that that issue corresponds to rows 9–12 and 25–28 in the demonstration above.

Also, the discussion below the demonstration noted that grouping by two levels that include missing values, even when `dropna=True`, retains the null groups, which seems incorrect and differs from the behavior of `.aggregate()`. That inconsistency is not specific to categorical groupers and was mentioned in a [comment](https://github.com/pandas-dev/pandas/issues/10484#issuecomment-117936025) to issue #10484. For example, the following `groupby` seems confused about its length:

```python
In [2]: gby = pd.Series(range(3)).groupby(2 * [pd.Series(["A", "B", pd.NA])])

In [3]: list(gby)
Out[3]: 
[(('A', 'A'),
  0    0
  dtype: int64),
 (('B', 'B'),
  1    1
  dtype: int64)]

In [4]: len(gby)
Out[4]: 3
```

The remaining related issue I see is the aforementioned issue #35202 that affects rows 1–4,

**Comment by jreback** [comment]
@stanwest i was asking if there is an issue for this comment: https://github.com/pandas-dev/pandas/pull/44180#issuecomment-995405764

**Comment by jreback** [comment]
thanks @stanwest

**Comment by rhshadrach** [comment]
@jreback - The issue @stanwest pointed to is the same as the one in my comment above. I think we're good here.

**Comment by stanwest** [comment]
Thank you for the discussion and for merging this.
